### ISSUE TITLE
support for retrying custom defined errors extended from Error class

### ISSUE THREAD

**Issue by danielamezcua** [vcfvct/typescript-retry-decorator#29]
### Use Case

I need to pass custom defined error classes on the `value` field  of the  `RetryOptions` options object:

### Code that reproduces error:
```
import { Retryable, BackOffPolicy} from "typescript-retry-decorator"

class CustomError extends Error {
  constructor(message?: string | undefined) {
    super(message)
    Object.setPrototypeOf(this, CustomError.prototype)
    this.name = 'CustomError'
  }
}

class Test {
   @Retryable({
    maxAttempts: 3,
    backOffPolicy: BackOffPolicy.FixedBackOffPolicy,
    backOff: 1000,
    value: [CustomError],
  })
    public async test() {
        throw new CustomError()
    }
}

```
### Error
```
Type 'typeof CustomError' is not assignable to type 'ErrorConstructor'.
  Type 'typeof CustomError' provides no match for the signature '(message?: string | undefined): Error'

```
 
My build target is `esnext` 

Could someone provide an example where a custom error class is passed to the `value` field of the `RetryOptions` object? The tests and examples do not include such case

**Comment by vcfvct** [comment]
I think this is because the `ErrorConstructor` interface expects a call signature that takes an optional message parameter of type string, and returns an instance of Error. However, your `CustomError` class does not have a call signature, it only has a constructor. This means that you cannot call `CustomError` as a function, like this:

> let error = CustomError("Some message"); // This will not work

You can only use the new keyword to create an instance of `CustomError`, like this:

> let error = new CustomError("Some message"); // This will work

This is different from the built-in error constructors, which can be used both as functions and as constructors. For example:
```
let error1 = SyntaxError("Some message"); // This will work
let error2 = new SyntaxError("Some message"); // This will also work
```

I guess the **easier** fix is to just use type assertion n to tell TypeScript that the error constructors are compatible with `ErrorConstructor`:
```
value: [CustomError as ErrorConstructor],

// or use any if needed:
value: [CustomError as any],
```

**Comment by danielamezcua** [comment]
@vcfvct thanks for the reply!
Based on your answer and after looking at Typescript documentation, it seems that there is no solution that does not feel cumbersome for this. 

Seems like by expecting an object that implements the `ErrorConstructor` interface, the Error classes/interfaces that can be becomes very limited. Is there any particular reason to have the `values` option receive only objects that implement the `ErrorConstructor` interface? I am curious on why this approach was preferred than expecting an `Error` class/subclass.

**Comment by vcfvct** [comment]
@danielamezcua , do you have any ideas of implement this type system here so that we can easily take `Error` and its subclass ? 
I took the `ErrorConstructor` just because it is the built-in Constructor for Error. When I get some time maybe I can also try if there's any other better options.

**Comment by danielamezcua** [comment]
@vcfvct  I guess that instead of expecting objects that implement the `ErrorConstructor` interface (which, as you said, require both a call signature and a constructor), the `value` field can expect any objects that have a constructor that return an `Error` instance, without necessarily having a call signature.

something like:

```
interface ConstructableError {
  new (...args: any[]): Error;
}

interface RetryOptions {
  ... 
  value?: ConstructableError[]
  ...
}
```
This would work with any built-in error interface like `SyntaxError`, `ReferenceError`, and with  any custom error class extended from the `Error` class.

**Comment by vcfvct** [comment]
coool man! @danielamezcua . I just created/merged a PR #29. and published a new version v2.4.0. 

Please let me know if it works. Thanks for bringing up the idea! 😄

**Comment by danielamezcua** [comment]
@vcfvct awesome! works great. thanks a lot
